From 593e728bef6a6548474a2f25d9a55902b5d65b2f Mon Sep 17 00:00:00 2001
From: HuaWen CHEN <quic_huawenc@quicinc.com>
Date: Mon, 18 Sep 2023 16:12:04 +0800
Subject: [PATCH] mwan3: change for default route

to support only one default route in main table:
1. default route change would be notified by named pipe.
2. ip monitor route handling, skip the default route change.
---
 net/mwan3/files/lib/mwan3/mwan3.sh  |  5 ++-
 net/mwan3/files/usr/sbin/mwan3rtmon | 57 ++++++++++++++++++++++++++---
 2 files changed, 56 insertions(+), 6 deletions(-)

diff --git a/net/mwan3/files/lib/mwan3/mwan3.sh b/net/mwan3/files/lib/mwan3/mwan3.sh
index 1bfb767e86..0ae2554197 100644
--- a/net/mwan3/files/lib/mwan3/mwan3.sh
+++ b/net/mwan3/files/lib/mwan3/mwan3.sh
@@ -455,13 +455,14 @@ mwan3_get_routes()
 {
 	{
 		$IP route list table main
+		cat $default_route_file
 		config_list_foreach "globals" "rt_table_lookup" mwan3_extra_tables_routes
 	} | sed -ne "$MWAN3_ROUTE_LINE_EXP" | sort -u
 }
 
 mwan3_create_iface_route()
 {
-	local tid route_line family IP id tbl
+	local tid route_line family IP id tbl default_route_file
 	config_get family "$1" family ipv4
 	mwan3_get_iface_id id "$1"
 
@@ -469,8 +470,10 @@ mwan3_create_iface_route()
 
 	if [ "$family" = "ipv4" ]; then
 		IP="$IP4"
+		default_route_file="/tmp/default.route.v4.list"
 	elif [ "$family" = "ipv6" ]; then
 		IP="$IP6"
+		default_route_file="/tmp/default.route.v6.list"
 	fi
 
 	tbl=$($IP route list table $id 2>/dev/null)$'\n'
diff --git a/net/mwan3/files/usr/sbin/mwan3rtmon b/net/mwan3/files/usr/sbin/mwan3rtmon
index b7f03cc874..c2f6074d88 100755
--- a/net/mwan3/files/usr/sbin/mwan3rtmon
+++ b/net/mwan3/files/usr/sbin/mwan3rtmon
@@ -69,10 +69,16 @@ mwan3_add_all_routes()
 mwan3_rtmon_route_handle()
 {
 	local action route_line family tbl device line tid
+	local isFromPipe=$3
 
 	route_line=${1##"Deleted "}
 	route_family=$2
 
+#	if [ "$isFromPipe" == "0" ] && [[ "$route_line" =~ ^default.* ]]; then
+#	  LOG warn "mwan3rtmon skip default route change in main table (default route maintained seperatly)"
+#	  return
+#	fi
+
 	if [ "$route_line" = "$1" ]; then
 		action="replace"
 		$IPS -! add mwan3_connected_${route_family##ip} ${route_line%% *}
@@ -135,9 +141,25 @@ mwan3_rtmon_route_handle()
 	fi
 }
 
+trap_with_arg_pid_list()
+{
+	func="$1" ; shift
+	sig="$1" ; shift
+	trap "$func $sig $*" "$sig"
+}
+
+func_trap_pidlist()
+{
+	sig=$1;shift
+
+	LOG warn "send kill $sig to pidlist: $*"
+	kill -$sig $* 2>/dev/null
+}
+
+
 main()
 {
-	local IP family
+	local IP family defaultRoutePipe
 
 	mwan3_init
 
@@ -149,9 +171,18 @@ main()
 			exit 1
 		fi
 		IP="$IP6"
+		defaultRoutePipe="/tmp/default.route.v6.pipe"
+		default_route_file="/tmp/default.route.v6.list"
 	else
 		IP="$IP4"
+		defaultRoutePipe="/tmp/default.route.v4.pipe"
+		default_route_file="/tmp/default.route.v4.list"
+	fi
+
+	if [ ! -f $defaultRoutePipe ]; then
+	  mkfifo $defaultRoutePipe
 	fi
+
 	sh -c "echo \$\$; exec $IP monitor route" | {
 		read -r monitor_pid
 		trap_with_arg func_trap "$monitor_pid" SIGINT SIGTERM SIGKILL
@@ -159,14 +190,30 @@ main()
 		while IFS='' read -r line; do
 			[ -z "${line##*table*}" ] && continue
 			LOG debug "handling route update $family '$line'"
-			mwan3_rtmon_route_handle "$line" "$family"
+			mwan3_rtmon_route_handle "$line" "$family" 0
 		done
 	} &
 	child=$!
-	trap_with_arg func_trap "$child" SIGINT SIGTERM SIGKILL
+
+	sh -c "echo \$\$; exec tail -f $defaultRoutePipe" | {
+		read -r monitor_pid
+		trap_with_arg func_trap "$monitor_pid" SIGINT SIGTERM SIGKILL
+		KILL -SIGSTOP $$
+		while IFS='' read -r line; do
+			[ -z "${line##*table*}" ] && continue
+			LOG debug "handling route update $family '$line'"
+			mwan3_rtmon_route_handle "$line" "$family" 1
+		done
+	} &
+	child_defroute=$!
+
+	trap_with_arg_pid_list func_trap_pidlist  SIGINT $child $child_defroute
+	trap_with_arg_pid_list func_trap_pidlist SIGTERM $child $child_defroute
+	trap_with_arg_pid_list func_trap_pidlist SIGKILL $child $child_defroute
+
 	mwan3_set_connected_${family}
 	mwan3_add_all_routes ${family}
-	kill -SIGCONT $child
-	wait $child
+	kill -SIGCONT $child $child_defroute
+	wait $child $child_defroute
 }
 main "$@"
-- 
2.40.0.windows.1

